libraries/AP_ESC_Telem/AP_ESC_Telem.cpp

```diff
@@ -35,23 +39,337 @@ AP_ESC_Telem::AP_ESC_Telem()
+        || !(_telem_data_mask[esc_index] & AP_ESC_Telem_Backend::TelemetryType::USAGE)) {
+        return false;
+    }
+    usage_s = _telem_data[esc_index].usage_s;
+    return true;
+}
+
+// send ESC telemetry messages over MAVLink
+void AP_ESC_Telem::send_esc_telemetry_mavlink(uint8_t mav_chan)
+{
+    // return immediately if no ESCs have been found
+    if (_active_escs_mask == 0) {
+        return;
+    }
+
+    static_assert(ESC_TELEM_MAX_ESCS <= 12, "AP_ESC_Telem::send_esc_telemetry_mavlink() only supports up-to 12 motors");
+
+    // loop through 3 groups of 4 ESCs
+    for (uint8_t i = 0; i < 3; i++) {
+
+        // return if no space in output buffer to send mavlink messages
+        if (!HAVE_PAYLOAD_SPACE((mavlink_channel_t)mav_chan, ESC_TELEMETRY_1_TO_4)) {
+            return;
+        }
+
+        // skip this group of ESCs if no data to send
+        if ((_active_escs_mask & ((uint32_t)0x0F << i*4)) == 0) {
+            continue;
+        }
+
+        // arrays to hold output
+        uint8_t temperature[4] {};
+        uint16_t voltage[4] {};
+        uint16_t current[4] {};
+        uint16_t current_tot[4] {};
+        uint16_t rpm[4] {};
+        uint16_t count[4] {};
+
+        // fill in output arrays
+        for (uint8_t j = 0; j < 4; j++) {
+            const uint8_t esc_id = i * 4 + j;
+            temperature[j] = _telem_data[esc_id].temperature_deg;
+            voltage[j] = _telem_data[esc_id].voltage_cv;
+            current[j] = _telem_data[esc_id].current_ca;
+            current_tot[j] = constrain_float(_telem_data[esc_id].consumption_mah, 0, UINT16_MAX);
+            float rpmf;
+            if (get_rpm(esc_id, rpmf)) {
+                rpm[j] = constrain_float(rpmf, 0, UINT16_MAX);
+            } else {
+                rpm[j] = 0;
+            }
+            count[j] = _telem_data[esc_id].count;
+        }
+
+        // send messages
+        switch (i) {
+            case 0:
+                mavlink_msg_esc_telemetry_1_to_4_send((mavlink_channel_t)mav_chan, temperature, voltage, current, current_tot, rpm, count);
+                break;
+            case 1:
+                mavlink_msg_esc_telemetry_5_to_8_send((mavlink_channel_t)mav_chan, temperature, voltage, current, current_tot, rpm, count);
+                break;
+            case 2:
+                mavlink_msg_esc_telemetry_9_to_12_send((mavlink_channel_t)mav_chan, temperature, voltage, current, current_tot, rpm, count);
+                break;
+            default:
+                break;
+        }
+    }
+}
+
+// record an update to the telemetry data together with timestamp
+// this should be called by backends when new telemetry values are available
+void AP_ESC_Telem::update_telem_data(uint8_t esc_index, const AP_ESC_Telem_Backend::TelemetryData& new_data, uint8_t data_mask)
+{
+    if (esc_index > ESC_TELEM_MAX_ESCS) {
+        return;
+    }
+
+    _last_telem_data_ms[esc_index] = AP_HAL::millis();
```

Comment:
i think we do need a WITH_SEMAPHORE()

Reply:
I've done this differently, I have factored all the data into a per-ESC data structure that are independent. We can also just get rid of "active escs" and just rely on the timestamp. I think this avoids the need for semaphores.